### PR TITLE
fix(spec): continue MCP cursor scan after deep siblings

### LINKED ISSUE #1306: MCP schema recursion can stack-overflow during source add
## Problem

A hostile MCP `tools/list` response can publish recursive JSON Schema that makes `coral source add` recurse until stack overflow. This is attacker-controlled input for MCP-backed DSL v4 materialization, so the importer needs bounded, cycle-aware schema walking.

## Evidence

- `crates/coral-spec/src/v4/surfaces/mcp/import.rs:410` inserts the `$ref` into `resolving_refs`.
- `crates/coral-spec/src/v4/surfaces/mcp/import.rs:422` removes that guard before `input_object_shape` recurses through `allOf` at `:364`.
- That means an `A -> allOf -> B -> allOf -> A` cycle is not held in the guard set for the full subtree.
- `find_response_cursor_path` recurses at `crates/coral-spec/src/v4/surfaces/mcp/import.rs:643` with no visited set or depth cap.

## Acceptance criteria

- Keep `$ref` guards active for the whole schema subtree, for example with an RAII-style guard whose `Drop` removes the key.
- Add a flat `MAX_SCHEMA_DEPTH` for MCP input schema walking and response cursor-path discovery.
- Return a diagnostic/import failure for cycles or excessive depth instead of stack overflow.
- Add regression tests for both `$ref`/`allOf` cycles and deeply nested cursor discovery.

### PR DESCRIPTION
**Pull request by simonwhitaker**
## Summary

Closes #1306.

- Keeps MCP input `$ref` guards active across the resolved schema subtree.
- Adds a bounded MCP schema depth limit for input walking and response cursor discovery.
- Continues response cursor discovery across sibling branches after one branch exceeds the depth limit.
- Emits diagnostics only when no cursor is found before the depth cap blocks the search.

## Validation

- `make rust-checks`

### CODE REVIEW

**Comment by simonwhitaker** [comment]
* **#1317** <a href="https://app.graphite.com/github/pr/withcoral/coral/1317?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1316** <a href="https://app.graphite.com/github/pr/withcoral/coral/1316?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1315** <a href="https://app.graphite.com/github/pr/withcoral/coral/1315?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/withcoral/coral/1315?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1314** <a href="https://app.graphite.com/github/pr/withcoral/coral/1314?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/coral-spec/src/v4/surfaces/mcp/input_schema.rs [hunk 1]
```diff
@@ -74,16 +80,31 @@ fn input_object_shape(
...
-    let schema = resolve_input_schema_ref(
-        root,
-        schema,
-        surface_id,
-        operation_id,
-        resolving_refs,
-        diagnostics,
-        schema_complete,
-    )?;
+    if depth > MAX_SCHEMA_DEPTH {
```
Comment (jsummerfield) [inline crates/coral-spec/src/v4/surfaces/mcp/input_schema.rs:85]: This guard is useful, but it does not fully close the MCP schema recursion hole. The `allOf` merge path can still reach `input_property_schemas_conflict`, which calls `schema_without_annotation_metadata()` and recursively descends arbitrary duplicate property schemas without any depth budget. A shallow `allOf` with the same property name in two branches and a deeply nested property schema can still overflow or burn CPU outside this new guard.

Please put the same budget on the schema-normalization/comparison path too, and add a regression test for duplicate `allOf` properties with a deeply nested schema.

File: crates/coral-spec/src/v4/surfaces/mcp/output_schema.rs [hunk 2]
```diff
@@ -148,8 +157,9 @@ fn wrapped_list_property(schema: &Value) -> Option<(&str, Option<&Value>)> {
         return None;
     }
     if properties.len() != 1
-        && find_response_cursor_path(schema).is_none()
         && !is_offset_pagination_envelope(properties)
+        && find_response_cursor_path_or_warn(schema, surface_id, operation_id, diagnostics)
```
Comment (jsummerfield) [inline crates/coral-spec/src/v4/surfaces/mcp/output_schema.rs]: This treats cursor-search `DepthExceeded` like “no cursor” while deciding whether the response is a wrapped list. That changes row extraction: a response with a shallow `items` array plus cursor metadata beyond `MAX_SCHEMA_DEPTH` now imports as `Singleton` with an empty `row_path`, so runtime extraction returns the whole envelope instead of the `items` rows.

The depth cap should disable/warn on cursor pagination inference, not discard an already-known wrapped-list row path.

**Review by jsummerfield — COMMENTED** [review]
Two actionable issues from the review. Both are local to the MCP schema recursion changes.
